# Mais sobre `SELECT `: agrupamento, subqueries

## Introdução

Vamos continuar nossa investigação sobre o comando `SELECT`. Desta vez usaremos a base de dados 'sakila', uma famosa base de dados de teste construída pelos autores do MySQL. 

Abra a URL [`https://dev.mysql.com/doc/sakila/en/`](https://dev.mysql.com/doc/sakila/en/) para conhecer melhor esta base de dados, que modela uma locadora de DVDs (riam, vocês também serão velhos um dia!). Se você não possui esta base, siga as instruções de instalação na página.

Vamos também construir nosso objeto auxiliar para conectar com a base de dados, como fizemos na última aula.

In [53]:
from functools import partial
import mysql.connector

connection = mysql.connector.connect(
    host='localhost',
    user='root',
    password='password',
    database='sakila',
    sql_mode='only_full_group_by',
)


def run_db_query(connection, query, args=None):
    with connection.cursor() as cursor:
        print('Executando query:')
        cursor.execute(query, args)
        for result in cursor:
            print(result)


db = partial(run_db_query, connection)

## Aquecimento

Quais os nomes das categorias de filme?

In [3]:
db('SELECT name FROM category')

Executando query:
('Action',)
('Animation',)
('Children',)
('Classics',)
('Comedy',)
('Documentary',)
('Drama',)
('Family',)
('Foreign',)
('Games',)
('Horror',)
('Music',)
('New',)
('Sci-Fi',)
('Sports',)
('Travel',)


Quais atores tem as iniciais "J.D."?

In [8]:
db('''
SELECT 
    first_name, last_name 
FROM 
    actor
WHERE
    (first_name LIKE 'J%') AND (last_name LIKE 'D%')
''')

Executando query:
('JENNIFER', 'DAVIS')
('JUDY', 'DEAN')
('JODIE', 'DEGENERES')
('JULIANNE', 'DENCH')


Liste as cidades brasileiras presentes na base de dados.

In [16]:
db('''
SELECT 
    city 
FROM 
    city
    INNER JOIN country USING (country_id)
WHERE
    COUNTRY = "Brazil"
''')

Executando query:
('Alvorada',)
('Angra dos Reis',)
('Anápolis',)
('Aparecida de Goiânia',)
('Araçatuba',)
('Bagé',)
('Belém',)
('Blumenau',)
('Boa Vista',)
('Brasília',)
('Goiânia',)
('Guarujá',)
('Águas Lindas de Goiás',)
('Ibirité',)
('Juazeiro do Norte',)
('Juiz de Fora',)
('Luziânia',)
('Maringá',)
('Poá',)
('Poços de Caldas',)
('Rio Claro',)
('Santa Brbara dOeste',)
('Santo André',)
('São Bernardo do Campo',)
('São Leopoldo',)
('Sorocaba',)
('Vila Velha',)
('Vitória de Santo Antão',)


*Quantas* cidades brasileiras tem na base de dados?

In [17]:
db('''
SELECT 
    count(city) 
FROM 
    city
    INNER JOIN country USING (country_id)
WHERE
    COUNTRY = "Brazil"
''')

Executando query:
(28,)


Liste os filmes do ator (fictício) "Dan Harris"

In [22]:
db('''
SELECT
    title
FROM
    film
    INNER JOIN film_actor USING (film_id)
    INNER JOIN actor USING (actor_id)
WHERE
    (first_name = "Dan") and (last_name = "Harris")
''')

Executando query:
('BEDAZZLED MARRIED',)
('BOONDOCK BALLROOM',)
('DESTINY SATURDAY',)
('DIVINE RESURRECTION',)
('EYES DRIVING',)
('FELLOWSHIP AUTUMN',)
('GHOST GROUNDHOG',)
('GROOVE FICTION',)
('HILLS NEIGHBORS',)
('HOLIDAY GAMES',)
('INDEPENDENCE HOTEL',)
('INSIDER ARIZONA',)
('JADE BUNCH',)
('LIES TREATMENT',)
('MONTEREY LABYRINTH',)
('REUNION WITCHES',)
('RUN PACIFIC',)
('SCHOOL JACKET',)
('SEVEN SWARM',)
('SIEGE MADRE',)
('STEERS ARMAGEDDON',)
('STRAIGHT HOURS',)
('SUMMER SCARFACE',)
('SUPERFLY TRIP',)
('TITANIC BOONDOCK',)
('TITANS JERK',)
('VANISHING ROCKY',)
('WATERSHIP FRONTIER',)


Quais filmes estão alugados por Florence Woods?

In [51]:
db('''
SELECT
    title
FROM
    customer
    INNER JOIN rental USING(customer_id)
    INNER JOIN inventory USING(inventory_id)
    INNER JOIN film USING(film_id)
WHERE
    (first_name = "Florence") and (last_name = "Woods")
''')

Executando query:
('SLEEPY JAPANESE',)
('CHARADE DUFFEL',)
('MIGHTY LUCK',)
('MASSACRE USUAL',)
('PEAK FOREVER',)
('SALUTE APOLLO',)
('WEEKEND PERSONAL',)
('MALLRATS UNITED',)
('RIVER OUTLAW',)
('PITY BOUND',)
('DEER VIRGINIAN',)
('KRAMER CHOCOLATE',)
('BINGO TALENTED',)
('RIDER CADDYSHACK',)
('HORROR REIGN',)
('GABLES METROPOLIS',)
('BROOKLYN DESERT',)
('GUYS FALCON',)
('SATURN NAME',)
('WARS PLUTO',)
('INDIAN LOVE',)
('SPINAL ROCKY',)
('ICE CROSSING',)
('PET HAUNTING',)
('CLUB GRAFFITI',)
('VANILLA DAY',)
('CINCINATTI WHISPERER',)
('SCALAWAG DUCK',)
('BLADE POLISH',)
('SHOOTIST SUPERFLY',)


In [ ]:
db('''
SELECT
    language.Name
    
FROM
    LANGUAGE
    LEFT OUTER JOIN FILM ON language.language_id = film.language_id
WHERE
    film.language_id IS NULL
''')

Para quais línguas não tem nenhum filme na locadora? (Dica: use `LEFT OUTER JOIN`.)

In [62]:
db('''
SELECT
    language.name
FROM
    language
    LEFT OUTER JOIN film ON language.language_id = film.language_id
WHERE
    film.language_id IS NULL
''')

Executando query:
('Italian',)
('Japanese',)
('Mandarin',)
('French',)
('German',)


## `DISTINCT`
As vezes desejamos consultar quais os valores distintos de uma coluna. Para isso usamos o qualificador `DISTINCT`. 

Por exemplo: Quais os anos de lançamento dos filmes da base? 

In [34]:
# Vai aparecer muitas cópias de '(2006,)'
db('SELECT release_year FROM film LIMIT 20')

Executando query:
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)
(2006,)


Ops, parece que não tem muita variedade nesta base! Usando `DISTINCT` podemos limpar esse resultado:

In [35]:
db('SELECT DISTINCT release_year FROM film')

Executando query:
(2006,)


### Praticando

Quais clientes estão alugando um DVD agora?

In [67]:
db('''
SELECT DISTINCT
    customer_id,first_name,last_name
FROM
    rental
    INNER JOIN customer USING(customer_id)
WHERE
    (active = 1) and (return_date IS NULL)
ORDER BY
    customer_id ASC
''')

Executando query:
(5, 'ELIZABETH', 'BROWN')
(9, 'MARGARET', 'MOORE')
(11, 'LISA', 'ANDERSON')
(14, 'BETTY', 'WHITE')
(15, 'HELEN', 'HARRIS')
(21, 'MICHELLE', 'CLARK')
(22, 'LAURA', 'RODRIGUEZ')
(23, 'SARAH', 'LEWIS')
(28, 'CYNTHIA', 'YOUNG')
(29, 'ANGELA', 'HERNANDEZ')
(33, 'ANNA', 'HILL')
(41, 'STEPHANIE', 'MITCHELL')
(42, 'CAROLYN', 'PEREZ')
(43, 'CHRISTINE', 'ROBERTS')
(44, 'MARIE', 'TURNER')
(52, 'JULIE', 'SANCHEZ')
(53, 'HEATHER', 'MORRIS')
(56, 'GLORIA', 'COOK')
(58, 'JEAN', 'BELL')
(60, 'MILDRED', 'BAILEY')
(69, 'JUDY', 'GRAY')
(73, 'BEVERLY', 'BROOKS')
(75, 'TAMMY', 'SANDERS')
(80, 'MARILYN', 'ROSS')
(83, 'LOUISE', 'JENKINS')
(87, 'WANDA', 'PATTERSON')
(91, 'LOIS', 'BUTLER')
(94, 'NORMA', 'GONZALES')
(99, 'EMILY', 'DIAZ')
(100, 'ROBIN', 'HAYES')
(101, 'PEGGY', 'MYERS')
(107, 'FLORENCE', 'WOODS')
(108, 'TRACY', 'COLE')
(111, 'CARMEN', 'OWENS')
(114, 'GRACE', 'ELLIS')
(115, 'WENDY', 'HARRISON')
(120, 'SYLVIA', 'ORTIZ')
(135, 'JUANITA', 'MASON')
(142, 'APRIL', 'BURNS')
(152, 'ALIC

## Agrupamento

Uma das características mais valiosas de banco de dados é o *agrupamento*. Podemos agrupar os resultados de uma query indicando uma coluna cujos valores serão usados para agrupar os dados.

Por exemplo, considere a seguinte tabela, que chamaremos de 'vendas':

| id | id_item | item | preco |
|--|--|--|--|
| 1 | 1 | A | 5 |
| 2 | 2 | B | 6 |
| 3 | 1 | A | 3 |
| 4 | 3 | C | 7 |
| 5 | 3 | C | 5 |
| 6 | 1 | A | 2 |

Se agruparmos pela coluna 'id_item' teremos 3 conjuntos de resultados:

id_item = 1:

| id | id_item | item | preco |
|--|--|--|--|
| 1 | 1 | A | 5 |
| 3 | 1 | A | 3 |
| 6 | 1 | A | 2 |

id_item = 2:

| id | id_item | item | preco |
|--|--|--|--|
| 2 | 2 | B | 6 |

id_item = 3:

| id | id_item | item | preco |
|--|--|--|--|
| 4 | 3 | C | 7 |
| 5 | 3 | C | 5 |

É como se tivessemos uma lista de tabelas! Isso não é permitido em SQL. Temos que 'resumir' a informação de cada uma das tabelas a uma linha só, o que significa que, para cada coluna, devemos escolher uma dessas opções:
- Resumir a informação da coluna usando uma função de grupo. Podemos somar, tirar a média, contar itens, concatená-los em uma única string, entre outras;
- Para colunas que se relacionam 1 para 1 com a coluna de agrupamento (como a coluna de agrupamento em si, ou a coluna 'item' neste exemplo), manter este valor. Isso acontece frequentemente quando fazemos `JOIN`.
- Não incluir a coluna, caso contrário.

Neste exemplo, podemos tomar a seguinte decisão para cada coluna:
- id: descartar
- id_item: manter valor
- item: manter valor
- preco: vamos calcular a soma dos valores, e renomear esta informação para 'total'

Com isso, obtemos a seguinte tabela:

| id_item | item | total |
|--|--|--|
| 1 | A | 10 |
| 2 | B | 6 |
| 3 | C | 12 |

Por fim, se não queremos id_item, ficamos com a seguinte tabela:

| item | total |
|--|--|
| A | 10 |
| B | 6 |
| C | 12 |

Para obter essa tabela podemos usar o seguinte comando SQL:

```SQL
SELECT 
    item, SUM(preco) as total 
FROM 
    vendas
GROUP BY
    id_item
```

Consulte o capítulo 9 do seu livro texto para conhecer mais sobre agrupamentos.

### Praticando

Quais os 10 atores que mais apareceram em filmes?

In [46]:
db('''
SELECT
    first_name,last_name, COUNT(actor_id) AS film_count
FROM
    actor
    INNER JOIN film_actor USING(actor_id)
GROUP BY
    actor_id
ORDER BY
    film_count DESC
LIMIT 10
''')

Executando query:
('GINA', 'DEGENERES', 42)
('WALTER', 'TORN', 41)
('MARY', 'KEITEL', 40)
('MATTHEW', 'CARREY', 39)
('SANDRA', 'KILMER', 37)
('SCARLETT', 'DAMON', 36)
('VAL', 'BOLGER', 35)
('VIVIEN', 'BASINGER', 35)
('GROUCHO', 'DUNST', 35)
('UMA', 'WOOD', 35)


## Pipeline do comando `SELECT`

Uma versão mais completa do `SELECT` (mas não inteiramente completa - consulte o manual do MySQL) é vista abaixo:

```mysql
SELECT [DISTINCT] <select_header> 
FROM <source_tables>
WHERE <filter_expression>
GROUP BY <grouping_expressions>
HAVING <filter_expression>
ORDER BY <ordering_expressions>
LIMIT <count> 
OFFSET <count>
```

Você já deve ter percebido que o comando `SELECT` tem uma sequência própria de avaliação. Por exemplo, para saber quais filmes custam mais que 3 dinheiros, podemos escrever:

In [47]:
db('''
SELECT 
    COUNT(f.rental_rate)
FROM
    film f
WHERE
    f.rental_rate > 3
''')

Executando query:
(336,)


Observe que o 'apelido' f para a tabela 'film' é definido na cláusula `FROM`, mas usado em `SELECT` e também em `WHERE`.

A ordem de execução do comando `SELECT` é aproximadamente como segue:

1. `FROM <source_tables>`: indica as tabelas que serão usadas nesta query e, conceitualmente, combina estas tabelas através de *produto cartesiano* em uma grande tabela. (Note o termo "*conceitualmente*" que usei: em termos de implementação da query este produto cartesiano raramente é construído.)

2. `WHERE <filter_expression>`: filtra linhas.

3. `GROUP BY <grouping_expressions>`: agrupa conjuntos de linhas.

4. `SELECT <select_heading>`: escolha de colunas e de agregados.

5. `HAVING <filter_expression>`: outra filtragem, esta aplicada apenas **depois** da agregação. Pode usar resultados do processo de agregação. Obriga o uso de `GROUP BY`.

6. `DISTINCT`: Elimina linhas duplicadas.

7. `ORDER BY`: ordena as linhas do resultado.

8. `OFFSET <count>`: Pula linhas do resultado. Requer LIMIT.

9. `LIMIT <count>`: Mantém apenas um número máximo de linhas.

Esta sequencia também serve como dica de como projetar uma query! 
- Comece identificando as tabelas que você deseja usar
- Monte o filtro de linhas, incluindo critérios de `JOIN`
- Agrupe
- Selecione colunas e aplique funções de agregação, conforme necessário
- Filtre com `HAVING`, agora que temos agregação
- O resto é mais fácil, aplique conforme requerido

## `WHERE` versus `HAVING`

Conforme visto acima, temos a cláusula `HAVING` para fazer filtragens *APÓS* agregação. Para que serve isso? Por exemplo, suponha que queremos saber quais atores não compartilham seu sobrenome com nenhum outro ator. Podemos usar a *query* a seguir:

In [48]:
db('''
SELECT 
    MIN(a.first_name) as fst_name, a.last_name
FROM 
    actor a 
GROUP BY 
    a.last_name
HAVING
    COUNT(a.first_name) = 1
ORDER BY
    a.last_name, fst_name
''')

Executando query:
('ANGELINA', 'ASTAIRE')
('RUSSELL', 'BACALL')
('HARRISON', 'BALE')
('RENEE', 'BALL')
('JULIA', 'BARRYMORE')
('VIVIEN', 'BASINGER')
('VIVIEN', 'BERGEN')
('LIZA', 'BERGMAN')
('CUBA', 'BIRCH')
('KEVIN', 'BLOOM')
('CHRIS', 'BRIDGES')
('LAURENCE', 'BULLOCK')
('MATTHEW', 'CARREY')
('GREG', 'CHAPLIN')
('RUSSELL', 'CLOSE')
('FRED', 'COSTNER')
('SIDNEY', 'CROWE')
('JUDE', 'CRUISE')
('RALPH', 'CRUZ')
('SCARLETT', 'DAMON')
('FRANCES', 'DAY-LEWIS')
('SYLVESTER', 'DERN')
('ALAN', 'DREYFUSS')
('GROUCHO', 'DUNST')
('CHRISTIAN', 'GABLE')
('MERYL', 'GIBSON')
('PARKER', 'GOLDBERG')
('ADAM', 'GRANT')
('MEG', 'HAWKE')
('GEOFFREY', 'HESTON')
('HARVEY', 'HOPE')
('ANGELA', 'HUDSON')
('CARMEN', 'HUNT')
('WHOOPI', 'HURT')
('WOODY', 'JOLIE')
('KIRK', 'JOVOVICH')
('MATTHEW', 'LEIGH')
('JOHNNY', 'LOLLOBRIGIDA')
('GRETA', 'MALDEN')
('ED', 'MANSFIELD')
('ELVIS', 'MARX')
('MORGAN', 'MCDORMAND')
('TOM', 'MIRANDA')
('BETTE', 'NICHOLSON')
('KENNETH', 'PESCI')
('OLYMPIA', 'PFEIFFER')
('GARY', 'PHOENIX'

## Praticando

Liste a duração média dos filmes na categoria 'Drama'. Aqui o `GROUP BY` é necessário?

Liste o nome da categoria e a duração média dos filmes por categoria.

Liste o nome da categoria e a duração média dos filmes por categoria, apenas para categorias cuja duração média de filme excede 120 minutos.

# Conclusão

Esta aula de hoje foi bastante densa! Dicas de estudo:

- Pratique no seu livro-texto, capítulo 9. Lembre-se que a base 'música' pode ser usada para praticar os comandos SQL vistos no livro.
- Tente criar queries que sirvam de exemplo para os conceitos do livro! A tarefa de criar exemplos é muito instrutiva!

**Leitura prévia**:
Para a próxima aula vamos continuar praticando, com os assuntos dos capítulos 10 e 11, prepare-se para a aula, ok?

Até a próxima!

In [ ]:
connection.close()